raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -191,29 +201,48 @@ public KafkaRaftClient(NetworkChannel channel,
         this.random = random;
         this.connections = new ConnectionCache(channel, bootstrapServers,
             retryBackoffMs, requestTimeoutMs, logContext);
-        this.unsentAppends = new ArrayBlockingQueue<>(10);
+        this.unwrittenAppends = new ArrayBlockingQueue<>(10);
+        this.uncommittedAppends = new TreeMap<>();
         this.connections.maybeUpdate(quorum.localId, new HostInfo(advertisedListener, bootTimestamp));
+        this.kafkaRaftMetrics = new KafkaRaftMetrics(metrics, "raft", quorum, bootTimestamp);
+        kafkaRaftMetrics.updateNumVoterConnections(1);
     }
 
     private void applyCommittedRecordsToStateMachine() {
         quorum.highWatermark().ifPresent(highWatermark -> {
             log.updateHighWatermark(highWatermark);
+            maybeCommitPendingAppends(highWatermark, time.milliseconds());
 
             logger.debug("Applying committed entries up to high watermark {}. " +
                 "Current position is {}", highWatermark, stateMachine.position());
 
             while (stateMachine.position().offset < highWatermark && shutdown.get() == null) {
                 OffsetAndEpoch position = stateMachine.position();
                 Records records = readCommitted(position, highWatermark);
-                logger.trace("Applying committed records at offset {} of size {} to the state machine",
-                    position, records.sizeInBytes());
+
                 stateMachine.apply(records);
                 logger.trace("Applied committed records at {} to the state machine; position " +
                     "updated to {}", position, stateMachine.position());
             }
         });
     }
 
+    private void maybeCommitPendingAppends(long highWatermark, long currentTimeMs) {
+        Iterator<Map.Entry<OffsetAndEpoch, BaseOffsetAndTime>> iter = uncommittedAppends.entrySet().iterator();
+        while (iter.hasNext()) {
+            Map.Entry<OffsetAndEpoch, BaseOffsetAndTime> entry = iter.next();
+            OffsetAndEpoch offsetAndEpoch = entry.getKey();
+            if (offsetAndEpoch.offset < highWatermark) {
+                BaseOffsetAndTime offsetAndTime = entry.getValue();
+                double elapsedTimePerRecord = (currentTimeMs - offsetAndTime.appendTimeMs) / (double) (offsetAndEpoch.offset - offsetAndTime.baseOffset + 1);
+                kafkaRaftMetrics.updateCommitLatency(elapsedTimePerRecord, currentTimeMs);
```

Comment:
We actually would not know the latency for each entry since the follower at the moment would only fetch at batch boundary, so we would not replicate part of a batch.